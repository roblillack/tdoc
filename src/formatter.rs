use crate::{Document, InlineStyle, Paragraph, ParagraphType, Span};
use std::collections::HashMap;
use std::io::Write;

const DEFAULT_WRAP_WIDTH: usize = 72;
const DEFAULT_QUOTE_PREFIX: &str = "| ";
const DEFAULT_UNORDERED_LIST_ITEM_PREFIX: &str = " â€¢ ";

#[derive(Clone)]
pub struct StyleTags {
    pub begin: String,
    pub end: String,
}

impl StyleTags {
    pub fn new(begin: impl Into<String>, end: impl Into<String>) -> Self {
        Self {
            begin: begin.into(),
            end: end.into(),
        }
    }
}

#[derive(Clone)]
pub struct FormattingStyle {
    pub reset_styles: String,
    pub text_styles: HashMap<InlineStyle, StyleTags>,
    pub quote_prefix: String,
    pub unordered_list_item_prefix: String,
    pub wrap_width: usize,
    pub left_padding: usize,
}

impl Default for FormattingStyle {
    fn default() -> Self {
        Self {
            reset_styles: String::new(),
            text_styles: HashMap::new(),
            quote_prefix: DEFAULT_QUOTE_PREFIX.to_string(),
            unordered_list_item_prefix: DEFAULT_UNORDERED_LIST_ITEM_PREFIX.to_string(),
            wrap_width: DEFAULT_WRAP_WIDTH,
            left_padding: 0,
        }
    }
}

impl FormattingStyle {
    pub fn ascii() -> Self {
        Self::default()
    }

    pub fn ansi() -> Self {
        let mut text_styles = HashMap::new();
        text_styles.insert(
            InlineStyle::Bold,
            StyleTags::new("\x1b[1m", "\x1b[22m"),
        );
        text_styles.insert(
            InlineStyle::Italic,
            StyleTags::new("\x1b[3m", "\x1b[23m"),
        );
        text_styles.insert(
            InlineStyle::Highlight,
            StyleTags::new("\x1b[7m", "\x1b[27m"),
        );
        text_styles.insert(
            InlineStyle::Underline,
            StyleTags::new("\x1b[4m", "\x1b[24m"),
        );
        text_styles.insert(
            InlineStyle::Strike,
            StyleTags::new("\x1b[9m", "\x1b[29m"),
        );

        Self {
            reset_styles: "\x1b[0m".to_string(),
            text_styles,
            quote_prefix: DEFAULT_QUOTE_PREFIX.to_string(),
            unordered_list_item_prefix: DEFAULT_UNORDERED_LIST_ITEM_PREFIX.to_string(),
            wrap_width: DEFAULT_WRAP_WIDTH,
            left_padding: 0,
        }
    }
}

pub struct Formatter<W: Write> {
    pub style: FormattingStyle,
    writer: W,
    current_line_width: usize,
}

impl<W: Write> Formatter<W> {
    pub fn new(writer: W, style: FormattingStyle) -> Self {
        Self {
            writer,
            style,
            current_line_width: 0,
        }
    }

    pub fn new_ascii(writer: W) -> Self {
        Self::new(writer, FormattingStyle::ascii())
    }

    pub fn new_ansi(writer: W) -> Self {
        Self::new(writer, FormattingStyle::ansi())
    }

    pub fn write_document(&mut self, document: &Document) -> std::io::Result<()> {
        let indent = " ".repeat(self.style.left_padding);
        self.write_paragraphs(&document.paragraphs, &indent, &indent)?;

        // Write reset styles if we have any
        if !self.style.reset_styles.is_empty() {
            write!(self.writer, "{}", self.style.reset_styles)?;
        }

        Ok(())
    }

    fn write_paragraphs(
        &mut self,
        paragraphs: &[Paragraph],
        prefix: &str,
        continuation_prefix: &str,
    ) -> std::io::Result<()> {
        for (i, paragraph) in paragraphs.iter().enumerate() {
            if i > 0 {
                writeln!(self.writer)?;
            }
            self.write_paragraph(paragraph, prefix, continuation_prefix)?;
        }
        Ok(())
    }

    fn write_paragraph(
        &mut self,
        paragraph: &Paragraph,
        prefix: &str,
        continuation_prefix: &str,
    ) -> std::io::Result<()> {
        match paragraph.paragraph_type {
            ParagraphType::Text
            | ParagraphType::Header1
            | ParagraphType::Header2
            | ParagraphType::Header3 => {
                self.write_text_paragraph(&paragraph.content, prefix, continuation_prefix)?;
            }
            ParagraphType::Quote => {
                let quote_prefix = format!("{}{}", prefix, self.style.quote_prefix);
                let quote_continuation = format!("{}{}", continuation_prefix, self.style.quote_prefix);
                
                for child in &paragraph.children {
                    self.write_paragraph(child, &quote_prefix, &quote_continuation)?;
                }
            }
            ParagraphType::UnorderedList => {
                for entry in &paragraph.entries {
                    let bullet_prefix = format!("{}{}", prefix, self.style.unordered_list_item_prefix);
                    let bullet_continuation = format!("{}  ", continuation_prefix);
                    
                    self.write_paragraphs(entry, &bullet_prefix, &bullet_continuation)?;
                }
            }
            ParagraphType::OrderedList => {
                for (i, entry) in paragraph.entries.iter().enumerate() {
                    let bullet_prefix = format!("{}{:2}. ", prefix, i + 1);
                    let bullet_continuation = format!("{}    ", continuation_prefix);
                    
                    self.write_paragraphs(entry, &bullet_prefix, &bullet_continuation)?;
                }
            }
        }
        Ok(())
    }

    fn write_text_paragraph(
        &mut self,
        spans: &[Span],
        prefix: &str,
        continuation_prefix: &str,
    ) -> std::io::Result<()> {
        if spans.is_empty() {
            return Ok(());
        }

        // Build the formatted text first
        let mut text_parts = Vec::new();
        for span in spans {
            self.collect_formatted_text(span, &mut text_parts)?;
        }

        // Now write with proper wrapping
        self.write_wrapped_text(&text_parts, prefix, continuation_prefix)?;
        writeln!(self.writer)?;

        Ok(())
    }

    fn collect_formatted_text(&self, span: &Span, parts: &mut Vec<String>) -> std::io::Result<()> {
        if span.children.is_empty() {
            // Handle line breaks specially
            if span.text.contains('\n') {
                for (i, line) in span.text.split('\n').enumerate() {
                    if i > 0 {
                        parts.push("\n".to_string());
                    }
                    if !line.is_empty() {
                        parts.push(line.to_string());
                    }
                }
            } else {
                parts.push(span.text.clone());
            }
        } else {
            // Apply styling to children
            if let Some(style_tags) = self.style.text_styles.get(&span.style) {
                parts.push(style_tags.begin.clone());
            }

            for child in &span.children {
                self.collect_formatted_text(child, parts)?;
            }

            if let Some(style_tags) = self.style.text_styles.get(&span.style) {
                parts.push(style_tags.end.clone());
            }
        }

        Ok(())
    }

    fn write_wrapped_text(
        &mut self,
        parts: &[String],
        prefix: &str,
        continuation_prefix: &str,
    ) -> std::io::Result<()> {
        // First, concatenate all non-newline parts to get the full text
        let mut full_text = String::new();
        let mut has_forced_breaks = false;
        
        for part in parts {
            if part == "\n" {
                has_forced_breaks = true;
                full_text.push('\n');
            } else {
                full_text.push_str(part);
            }
        }

        // If we have forced line breaks, handle them specially
        if has_forced_breaks {
            let lines: Vec<&str> = full_text.split('\n').collect();
            for (i, line) in lines.iter().enumerate() {
                if i > 0 {
                    writeln!(self.writer)?;
                    write!(self.writer, "{}", continuation_prefix)?;
                }
                if i == 0 {
                    write!(self.writer, "{}", prefix)?;
                }
                self.write_wrapped_line(line, continuation_prefix)?;
            }
        } else {
            write!(self.writer, "{}", prefix)?;
            self.write_wrapped_line(&full_text, continuation_prefix)?;
        }

        Ok(())
    }

    fn write_wrapped_line(&mut self, text: &str, continuation_prefix: &str) -> std::io::Result<()> {
        if text.is_empty() {
            return Ok(());
        }

        let words: Vec<&str> = text.split_whitespace().collect();
        let mut current_line = String::new();
        let mut line_width = 0;

        for (i, word) in words.iter().enumerate() {
            let word_width = self.visible_width(word);
            let space_needed = if i > 0 { 1 } else { 0 };
            
            // Check if we need to wrap
            if line_width + space_needed + word_width > self.style.wrap_width && !current_line.is_empty() {
                // Write current line and start a new one
                write!(self.writer, "{}", current_line.trim_end())?;
                writeln!(self.writer)?;
                write!(self.writer, "{}", continuation_prefix)?;
                current_line.clear();
                line_width = continuation_prefix.chars().count();
            }
            
            // Add space if needed
            if !current_line.is_empty() {
                current_line.push(' ');
                line_width += 1;
            }
            
            current_line.push_str(word);
            line_width += word_width;
        }

        // Write any remaining content
        if !current_line.is_empty() {
            write!(self.writer, "{}", current_line)?;
        }

        Ok(())
    }

    fn visible_width(&self, text: &str) -> usize {
        // Remove ANSI escape sequences for width calculation
        let ansi_regex = regex::Regex::new(r"\x1b\[[0-9;]*m").unwrap();
        let visible_text = ansi_regex.replace_all(text, "");
        visible_text.chars().count()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_helpers::*;

    #[test]
    fn test_ascii_formatting() {
        let mut output = Vec::new();
        let mut formatter = Formatter::new_ascii(&mut output);
        
        let doc = doc(vec![p_(vec![
            span("Hello "),
            b__("world"),
            span("!"),
        ])]);

        formatter.write_document(&doc).unwrap();
        let result = String::from_utf8(output).unwrap();
        
        println!("ASCII format result: '{}'", result);
        
        // ASCII formatter should not add any styling
        assert!(result.contains("Hello world!"));
        assert!(!result.contains("\x1b["));
    }

    #[test]
    fn test_ansi_formatting() {
        let mut output = Vec::new();
        let mut formatter = Formatter::new_ansi(&mut output);
        
        let doc = doc(vec![p_(vec![
            span("Hello "),
            b__("world"),
            span("!"),
        ])]);

        formatter.write_document(&doc).unwrap();
        let result = String::from_utf8(output).unwrap();
        
        // ANSI formatter should add bold styling
        assert!(result.contains("\x1b[1m")); // Bold begin
        assert!(result.contains("\x1b[22m")); // Bold end
        assert!(result.contains("\x1b[0m")); // Reset at end
    }

    #[test]
    fn test_quote_formatting() {
        let mut output = Vec::new();
        let mut formatter = Formatter::new_ascii(&mut output);
        
        let doc = doc(vec![quote_(vec![p__("Quoted text")])]);

        formatter.write_document(&doc).unwrap();
        let result = String::from_utf8(output).unwrap();
        
        assert!(result.contains("| Quoted text"));
    }

    #[test]
    fn test_list_formatting() {
        let mut output = Vec::new();
        let mut formatter = Formatter::new_ascii(&mut output);
        
        let doc = doc(vec![ul_(vec![
            li_(vec![p__("Item 1")]),
            li_(vec![p__("Item 2")]),
        ])]);

        formatter.write_document(&doc).unwrap();
        let result = String::from_utf8(output).unwrap();
        
        assert!(result.contains(" â€¢ Item 1"));
        assert!(result.contains(" â€¢ Item 2"));
    }

    #[test]
    fn test_wrap_width() {
        let mut output = Vec::new();
        let mut style = FormattingStyle::ascii();
        style.wrap_width = 20; // Very short for testing
        let mut formatter = Formatter::new(&mut output, style);
        
        let doc = doc(vec![p__("This is a very long line that should definitely be wrapped")]);

        formatter.write_document(&doc).unwrap();
        let result = String::from_utf8(output).unwrap();
        
        // Should contain line breaks due to wrapping
        let lines: Vec<&str> = result.lines().collect();
        assert!(lines.len() > 1);
    }
}