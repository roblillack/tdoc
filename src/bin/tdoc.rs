use clap::{Parser, ValueEnum, ValueHint};
use crossterm::terminal;
use reqwest::blocking::Client;
use std::fs::File;
use std::io::{self, BufReader, BufWriter, Read, Write};
use std::path::{Path, PathBuf};
use std::process::{Command as Process, Stdio};
use std::time::Duration;
use tdoc::formatter::{Formatter, FormattingStyle};
use tdoc::{html, markdown, pager, parse, write, Document};
use url::Url;

#[derive(Parser)]
#[command(
    name = "tdoc",
    version,
    about = "View and export FTML, HTML, and Markdown documents"
)]
struct Cli {
    /// Input file or URL (omit to read from stdin)
    #[arg(value_name = "INPUT")]
    input: Option<String>,

    /// Disable ANSI escape sequences in terminal output
    #[arg(long = "no-ansi")]
    no_ansi: bool,

    /// Explicitly set the input format when auto-detection is insufficient
    #[arg(long = "input-format", value_enum)]
    input_format: Option<InputFormatArg>,

    /// Write the rendered document to a file instead of the terminal
    #[arg(short = 'o', long = "output", value_name = "FILE", value_hint = ValueHint::FilePath)]
    output: Option<PathBuf>,
}

#[derive(Copy, Clone, Debug, Eq, PartialEq)]
enum InputFormat {
    Ftml,
    Html,
    Markdown,
}

#[derive(Copy, Clone, ValueEnum)]
enum InputFormatArg {
    Ftml,
    Html,
    Markdown,
}

impl From<InputFormatArg> for InputFormat {
    fn from(value: InputFormatArg) -> Self {
        match value {
            InputFormatArg::Ftml => InputFormat::Ftml,
            InputFormatArg::Html => InputFormat::Html,
            InputFormatArg::Markdown => InputFormat::Markdown,
        }
    }
}

struct InputSource {
    format: InputFormat,
    reader: Box<dyn Read>,
    display_name: String,
}

enum OutputFormat {
    Text,
    Ftml,
    Markdown,
    Html,
}

fn main() {
    if let Err(message) = run() {
        eprintln!("{message}");
        std::process::exit(1);
    }
}

fn run() -> Result<(), String> {
    let cli = Cli::parse();
    let input_override = cli.input_format.map(InputFormat::from);
    let input_source = create_reader(cli.input.as_deref(), input_override)?;
    let InputSource {
        format,
        reader,
        display_name,
    } = input_source;
    let document = parse_document(format, reader, &display_name)?;

    if let Some(output_path) = cli.output {
        write_output(&document, &output_path)?;
    } else {
        view_document(&document, cli.no_ansi)?;
    }

    Ok(())
}

fn create_reader(
    argument: Option<&str>,
    override_format: Option<InputFormat>,
) -> Result<InputSource, String> {
    match argument {
        None => Ok(InputSource {
            format: override_format.unwrap_or(InputFormat::Ftml),
            reader: Box::new(io::stdin()),
            display_name: "stdin".to_string(),
        }),
        Some("-") => Ok(InputSource {
            format: override_format.unwrap_or(InputFormat::Ftml),
            reader: Box::new(io::stdin()),
            display_name: "stdin".to_string(),
        }),
        Some(value) => {
            if let Ok(url) = Url::parse(value) {
                if url.scheme() == "http" || url.scheme() == "https" {
                    let client = Client::builder()
                        .timeout(Duration::from_secs(10))
                        .build()
                        .map_err(|err| format!("Unable to initialize HTTP client: {err}"))?;
                    let response = client
                        .get(value)
                        .send()
                        .map_err(|err| format!("Unable to fetch {value}: {err}"))?;
                    let extension = Path::new(url.path())
                        .extension()
                        .and_then(|ext| ext.to_str());
                    let format = override_format
                        .or_else(|| detect_input_format(extension))
                        .unwrap_or(InputFormat::Html);
                    return Ok(InputSource {
                        format,
                        reader: Box::new(response),
                        display_name: value.to_string(),
                    });
                }
            }

            let path = Path::new(value);
            let file = File::open(path)
                .map_err(|err| format!("Unable to open {value} for reading: {err}"))?;
            let extension = path.extension().and_then(|ext| ext.to_str());
            let format = override_format
                .or_else(|| detect_input_format(extension))
                .unwrap_or(InputFormat::Ftml);

            Ok(InputSource {
                format,
                reader: Box::new(BufReader::new(file)),
                display_name: value.to_string(),
            })
        }
    }
}

fn detect_input_format(extension: Option<&str>) -> Option<InputFormat> {
    let ext = extension?.to_ascii_lowercase();
    match ext.as_str() {
        "ftml" => Some(InputFormat::Ftml),
        "html" | "htm" => Some(InputFormat::Html),
        "md" | "markdown" => Some(InputFormat::Markdown),
        _ => None,
    }
}

fn parse_document(
    format: InputFormat,
    reader: Box<dyn Read>,
    display_name: &str,
) -> Result<Document, String> {
    match format {
        InputFormat::Ftml => {
            parse(reader).map_err(|err| format!("Unable to parse {display_name} as FTML: {err}"))
        }
        InputFormat::Html => html::parse(reader)
            .map_err(|err| format!("Unable to parse {display_name} as HTML: {err}")),
        InputFormat::Markdown => markdown::parse(reader)
            .map_err(|err| format!("Unable to parse {display_name} as Markdown: {err}")),
    }
}

fn view_document(document: &Document, no_ansi: bool) -> Result<(), String> {
    let stdout_is_tty = atty::is(atty::Stream::Stdout);
    let use_ansi = !no_ansi && stdout_is_tty;
    let use_pager = use_ansi;

    if use_pager {
        // if let Ok((mut pager, pager_stdin)) = run_pager() {
        //     let mut style = FormattingStyle::ansi();
        //     configure_style_for_terminal(&mut style);

        //     let mut formatter = Formatter::new(pager_stdin, style);
        //     formatter
        //         .write_document(document)
        //         .map_err(|err| format!("Unable to write document: {err}"))?;

        //     drop(formatter);
        //     let _ = pager.wait();
        //     return Ok(());
        // }

        let mut buf = Vec::new();
        let mut formatter = if use_ansi {
            let mut style = FormattingStyle::ansi();
            configure_style_for_terminal(&mut style);
            Formatter::new(&mut buf, style)
        } else {
            Formatter::new_ascii(&mut buf)
        };

        formatter
            .write_document(document)
            .map_err(|err| format!("Unable to write document: {err}"))?;

        let s = String::from_utf8(buf).map_err(|err| format!("UTF-8 error: {err}"))?;
        return pager::page_output(&s);
    }

    let mut formatter = if use_ansi {
        let mut style = FormattingStyle::ansi();
        configure_style_for_terminal(&mut style);
        Formatter::new(io::stdout(), style)
    } else {
        Formatter::new_ascii(io::stdout())
    };

    formatter
        .write_document(document)
        .map_err(|err| format!("Unable to write document: {err}"))
}

fn configure_style_for_terminal(style: &mut FormattingStyle) {
    if let Ok((width, _height)) = terminal::size() {
        let width = width as usize;
        if width < 60 {
            style.wrap_width = width;
            style.left_padding = 0;
        } else if width < 100 {
            style.wrap_width = width.saturating_sub(2);
            style.left_padding = 2;
        } else {
            let padding = (width.saturating_sub(100)) / 2 + 4;
            style.wrap_width = width.saturating_sub(padding);
            style.left_padding = padding;
        }
    }
}

fn run_pager() -> io::Result<(std::process::Child, std::process::ChildStdin)> {
    let pager_cmd = std::env::var("PAGER").unwrap_or_else(|_| "less".to_string());
    let mut parts: Vec<&str> = pager_cmd.split_whitespace().collect();

    if parts.is_empty() {
        parts.push("less");
    }

    let program = parts[0];
    let program_name = Path::new(program)
        .file_name()
        .and_then(|name| name.to_str())
        .unwrap_or(program);

    let mut final_args: Vec<&str> = parts.into_iter().skip(1).collect();
    if program_name == "less" || program_name == "more" {
        final_args.push("-R");
    }

    let mut child = Process::new(program)
        .args(&final_args)
        .stdin(Stdio::piped())
        .stdout(Stdio::inherit())
        .spawn()?;

    let stdin = child
        .stdin
        .take()
        .ok_or_else(|| io::Error::other("Unable to open pager stdin"))?;

    Ok((child, stdin))
}

fn write_output(document: &Document, output_path: &Path) -> Result<(), String> {
    if output_path == Path::new("-") {
        return Err(
            "Use stdout by omitting --output; it already writes to stdout by default.".to_string(),
        );
    }

    let extension = output_path
        .extension()
        .and_then(|ext| ext.to_str())
        .map(|ext| ext.to_ascii_lowercase());

    let format = determine_output_format(extension.as_deref()).ok_or_else(|| {
        format!(
            "Unable to determine output format for {}",
            output_path.display()
        )
    })?;

    match format {
        OutputFormat::Text => {
            let file = File::create(output_path).map_err(|err| {
                format!(
                    "Unable to open {} for writing: {err}",
                    output_path.display()
                )
            })?;
            let mut formatter = Formatter::new_ascii(file);
            formatter.write_document(document).map_err(|err| {
                format!(
                    "Unable to write document to {}: {err}",
                    output_path.display()
                )
            })?;
            Ok(())
        }
        OutputFormat::Ftml => {
            let mut file = File::create(output_path).map_err(|err| {
                format!(
                    "Unable to open {} for writing: {err}",
                    output_path.display()
                )
            })?;
            write(&mut file, document).map_err(|err| {
                format!("Unable to write FTML to {}: {err}", output_path.display())
            })?;
            file.flush()
                .map_err(|err| format!("Unable to flush {}: {err}", output_path.display()))
        }
        OutputFormat::Markdown => {
            let mut file = File::create(output_path).map_err(|err| {
                format!(
                    "Unable to open {} for writing: {err}",
                    output_path.display()
                )
            })?;
            markdown::write(&mut file, document).map_err(|err| {
                format!(
                    "Unable to write Markdown to {}: {err}",
                    output_path.display()
                )
            })?;
            file.flush()
                .map_err(|err| format!("Unable to flush {}: {err}", output_path.display()))
        }
        OutputFormat::Html => {
            let mut file = File::create(output_path).map_err(|err| {
                format!(
                    "Unable to open {} for writing: {err}",
                    output_path.display()
                )
            })?;
            write_html_document(&mut file, document).map_err(|err| {
                format!("Unable to write HTML to {}: {err}", output_path.display())
            })?;
            file.flush()
                .map_err(|err| format!("Unable to flush {}: {err}", output_path.display()))
        }
    }
}

fn determine_output_format(extension: Option<&str>) -> Option<OutputFormat> {
    let ext = extension?;
    match ext {
        "txt" | "text" => Some(OutputFormat::Text),
        "ftml" => Some(OutputFormat::Ftml),
        "md" | "markdown" => Some(OutputFormat::Markdown),
        "html" | "htm" => Some(OutputFormat::Html),
        _ => None,
    }
}

fn write_html_document<W: Write>(mut writer: W, document: &Document) -> io::Result<()> {
    writer.write_all(
        b"<!DOCTYPE html>\n<html>\n<head>\n<meta charset=\"utf-8\" />\n</head>\n<body>\n",
    )?;
    write(&mut writer, document)?;
    writer.write_all(b"\n</body>\n</html>\n")
}
