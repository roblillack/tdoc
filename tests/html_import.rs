use std::fs::File;
use std::io::Write;
use std::path::PathBuf;
use std::{fs, vec};

use insta::Settings;
use tdoc::formatter::Formatter;
use tdoc::{html, write};

fn collect_html_fixtures() -> Vec<PathBuf> {
    let data_dir = PathBuf::from("tests/data/html");
    let mut fixtures = Vec::new();

    if let Ok(entries) = fs::read_dir(&data_dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path
                .extension()
                .and_then(|ext| ext.to_str())
                .map_or(false, |ext| ext.eq_ignore_ascii_case("html"))
            {
                fixtures.push(path);
            }
        }
    }

    fixtures.sort();
    fixtures
}

#[test]
fn html_import_snapshots() {
    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_path("snapshots/html/import");
    settings.set_prepend_module_to_snapshot(false);
    settings.bind(|| {
        let fixtures = collect_html_fixtures();
        assert!(
            !fixtures.is_empty(),
            "expected HTML fixtures under tests/data/html"
        );

        for html_path in fixtures {
            let document = {
                let file = File::open(&html_path).expect("unable to open HTML fixture");
                html::parse(file).expect("failed to parse HTML fixture")
            };

            let mut rendered: Vec<u8> = vec![];
            Formatter::new_ascii(&mut rendered)
                .write_document(&document)
                .expect("failed to format document");
            // write(&mut rendered, &document).expect("failed to write FTML snapshot");

            let snapshot_name = format!(
                "{}.txt",
                html_path
                    .file_stem()
                    .and_then(|stem| stem.to_str())
                    .unwrap_or_default()
            );

            insta::assert_binary_snapshot!(snapshot_name.as_str(), rendered);
        }
    });
}
