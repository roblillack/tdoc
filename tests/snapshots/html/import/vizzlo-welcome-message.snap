---
source: tests/html_import.rs
expression: rendered
extension: txt
snapshot_kind: binary
---
